### PR TITLE
Feature/roc 4878 store document url on ccd

### PR DESCRIPTION
**Pull request by johnhosolirius**
**Before creating a pull request make sure that:**

- [ ] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [ ] tests have been updated / new tests has been added (if needed)

Please remove this line and everything above and fill the following sections:

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ X] No
```

### CODE REVIEW

File: ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/mapper/CaseEventMapper.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+        claimDocumentTypeMap.put(SEALED_CLAIM, SEALED_CLAIM_UPLOAD);
+        claimDocumentTypeMap.put(CLAIM_ISSUE_RECEIPT, CLAIM_ISSUE_RECEIPT_UPLOAD);
+        claimDocumentTypeMap.put(DEFENDANT_RESPONSE_RECEIPT, DEFENDANT_RESPONSE_UPLOAD);
+        claimDocumentTypeMap.put(CCJ_REQUEST, CCJ_REQUEST_UPLOAD);
+        claimDocumentTypeMap.put(SETTLEMENT_AGREEMENT, SETTLEMENT_AGREEMENT_UPLOAD);
+        claimDocumentTypeMap.put(DEFENDANT_PIN_LETTER, DEFENDANT_PIN_LETTER_UPLOAD);
+    }
+
+    public static CaseEvent map(ClaimDocumentType type) {
+
```
Comment (karthikkannankainos) [inline ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/mapper/CaseEventMapper.java]: Dont use default. if the type is not there, throw exception.

File: ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/mapper/CaseEventMapper.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+import java.util.Map;
+
+import static uk.gov.hmcts.cmc.ccd.domain.CaseEvent.*;
+import static uk.gov.hmcts.cmc.domain.models.ClaimDocumentType.*;
+
+public class CaseEventMapper {
+
+    private static final Map<ClaimDocumentType, CaseEvent> claimDocumentTypeMap;
+
+    static {
```
Comment (karthikkannankainos) [inline ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/mapper/CaseEventMapper.java]: Please remove the static block. just add these while declaration like new Hashmap(){  put();    }

File: ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/mapper/ClaimDocumentCollectionMapper.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+                .stream()
+                .map(claimDocumentMapper::to)
+                .collect(Collectors.toList())
+        );
+    }
+
+    public void from(CCDCase ccdCase, Claim.ClaimBuilder builder) {
+        Objects.requireNonNull(ccdCase, "ccdCase must not be null");
+
+        if (ccdCase.getCaseDocuments() == null || ccdCase.getCaseDocuments().isEmpty()) {
```
Comment (karthikkannankainos) [inline ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/mapper/ClaimDocumentCollectionMapper.java]: CollectionUtils.isEmpty in apache will do both empty and null check.

File: ccd-adapter/src/test/java/uk/gov/hmcts/cmc/ccd/mapper/ClaimDocumentMapperTest.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+@ContextConfiguration(classes = CCDAdapterConfig.class)
+@RunWith(SpringJUnit4ClassRunner.class)
+public class ClaimDocumentMapperTest {
+
+    @Autowired
+    private ClaimDocumentMapper claimDocumentMapper;
+
+    private ClaimDocument claimDocument;
+
+    private static final String NAME = "foo";
```
Comment (karthikkannankainos) [inline ccd-adapter/src/test/java/uk/gov/hmcts/cmc/ccd/mapper/ClaimDocumentMapperTest.java]: do we need these to be static?

File: ccd-adapter/src/test/java/uk/gov/hmcts/cmc/ccd/mapper/ClaimDocumentMapperTest.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+            .authoredDatetime(now)
+            .createdDatetime(yesterday)
+            .createdBy(CREATED_BY)
+            .build();
+    }
+
+    @Test
+    public void mapTo() {
+        CCDCollectionElement<CCDClaimDocument> ccdCollectionElement = claimDocumentMapper.to(claimDocument);
+
```
Comment (karthikkannankainos) [inline ccd-adapter/src/test/java/uk/gov/hmcts/cmc/ccd/mapper/ClaimDocumentMapperTest.java]: ccdCollectionElement.getValue() can be assigned to a variable and used for check every assert.

File: ccd-adapter/src/test/java/uk/gov/hmcts/cmc/ccd/mapper/ClaimDocumentMapperTest.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+            URI.create(ccdCollectionElement.getValue().getDocumentLink().getDocumentUrl()));
+        assertEquals(CCDClaimDocumentType.CLAIM_ISSUE_RECEIPT, ccdCollectionElement.getValue().getDocumentType());
+        assertEquals(claimDocument.getAuthoredDatetime(), ccdCollectionElement.getValue().getAuthoredDatetime());
+        assertEquals(claimDocument.getCreatedDatetime(), ccdCollectionElement.getValue().getCreatedDatetime());
+        assertEquals(claimDocument.getCreatedBy(), ccdCollectionElement.getValue().getCreatedBy());
+    }
+
+    @Test
+    public void mapFrom() {
+        CCDCollectionElement<CCDClaimDocument> ccdCollectionElement = claimDocumentMapper.to(claimDocument);
```
Comment (karthikkannankainos) [inline ccd-adapter/src/test/java/uk/gov/hmcts/cmc/ccd/mapper/ClaimDocumentMapperTest.java]: from() method to be tested not the to

File: gradle/wrapper/gradle-wrapper.properties [hunk 7]
```diff
@@ -1,5 +1,5 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-4.10-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-5.0-all.zip
```
Comment (karthikkannankainos) [inline gradle/wrapper/gradle-wrapper.properties]: remove this change. should go as part of separate PR.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/events/DocumentUploadHandler.java [hunk 8]
```diff
@@ -62,9 +63,7 @@ public void uploadDocument(RepresentedClaimIssuedEvent event) {
     public void uploadDocument(DefendantResponseEvent event) {
         Claim claim = event.getClaim();
         requireNonNull(claim, CLAIM_MUST_NOT_BE_NULL);
-        if (!claim.getResponse().isPresent()) {
-            throw new IllegalArgumentException("Response must be present");
-        }
+        claim.getResponse().orElseThrow(() -> new IllegalArgumentException("Response must be present"));
```
Comment (karthikkannankainos) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/events/DocumentUploadHandler.java]: IllegalstateException or NotFoundException. Pls talk to @srineelesh to be inline with what he agreed.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/events/ccd/CCDCaseHandler.java [hunk 9]
```diff
@@ -21,16 +20,11 @@
 import java.util.function.Predicate;
 
 import static org.springframework.transaction.event.TransactionPhase.BEFORE_COMMIT;
-import static uk.gov.hmcts.cmc.ccd.domain.CaseEvent.INTERLOCUTORY_JUDGMENT;
-import static uk.gov.hmcts.cmc.ccd.domain.CaseEvent.REJECT_ORGANISATION_PAYMENT_PLAN;
-import static uk.gov.hmcts.cmc.ccd.domain.CaseEvent.SETTLED_PRE_JUDGMENT;
-import static uk.gov.hmcts.cmc.claimstore.appinsights.AppInsights.CCD_LINK_DEFENDANT_ID;
-import static uk.gov.hmcts.cmc.claimstore.appinsights.AppInsights.CLAIM_EXTERNAL_ID;
-import static uk.gov.hmcts.cmc.claimstore.appinsights.AppInsights.REFERENCE_NUMBER;
+import static uk.gov.hmcts.cmc.ccd.domain.CaseEvent.*;
```
Comment (karthikkannankainos) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/events/ccd/CCDCaseHandler.java]: no star imports pls

File: src/main/java/uk/gov/hmcts/cmc/claimstore/repositories/CCDCaseRepository.java [hunk 10]
```diff
@@ -5,11 +5,7 @@
 import org.springframework.stereotype.Service;
 import uk.gov.hmcts.cmc.ccd.domain.CaseEvent;
 import uk.gov.hmcts.cmc.claimstore.services.ccd.CoreCaseDataService;
-import uk.gov.hmcts.cmc.domain.models.Claim;
-import uk.gov.hmcts.cmc.domain.models.ClaimDocumentCollection;
-import uk.gov.hmcts.cmc.domain.models.CountyCourtJudgment;
-import uk.gov.hmcts.cmc.domain.models.PaidInFull;
-import uk.gov.hmcts.cmc.domain.models.ReDetermination;
+import uk.gov.hmcts.cmc.domain.models.*;
```
Comment (karthikkannankainos) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/repositories/CCDCaseRepository.java]: no star import pls

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/ClaimService.java [hunk 11]
```diff
@@ -228,6 +223,8 @@ public Claim saveClaim(
                     new NotFoundException("Could not find claim with external ID '" + externalId + "'"));
         }
 
+        ccdEventProducer.createCCDClaimIssuedEvent(issuedClaim, authorisation);
```
Comment (karthikkannankainos) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/ClaimService.java]: any reason why this is moved up ?

File: src/main/java/uk/gov/hmcts/cmc/claimstore/events/DocumentUploadHandler.java (hunk 8 above)
Comment (srineelesh) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/events/DocumentUploadHandler.java]: @karthikkannankainos  : I am fixing this in my PR as this is an oversight on my part.

File: ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/domain/CCDClaimDocumentType.java [hunk 12]
```diff
@@ -0,0 +1,21 @@
+package uk.gov.hmcts.cmc.ccd.domain;
+
+public enum CCDClaimDocumentType {
+
+    CLAIM_ISSUE_RECEIPT("ClaimReceiptUpload"),
```
[the remark this answers is not in the document]
Comment (johnhosolirius) [reply ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/domain/CCDClaimDocumentType.java]: Yep. The description is not used, so a mistake.

File: ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/domain/CCDClaimDocumentType.java [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+    SEALED_CLAIM("SealedClaimUpload"),
+    DEFENDANT_RESPONSE_RECEIPT("DefendantResponseUpload"),
+    CCJ_REQUEST("CcjRequestUpload"),
+    SETTLEMENT_AGREEMENT("SettlementAgreementUpload"),
+    DEFENDANT_PIN_LETTER("DefendantPinLetterUpload");
+
+    private String typeName;
```
[the remark this answers is not in the document]
Comment (johnhosolirius) [reply ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/domain/CCDClaimDocumentType.java]: Resolved as above comment.

File: ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/mapper/CaseEventMapper.java (hunk 2 above)
Comment (johnhosolirius) [reply ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/mapper/CaseEventMapper.java]: Done

File: ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/mapper/ClaimDocumentCollectionMapper.java (hunk 3 above)
Comment (johnhosolirius) [reply ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/mapper/ClaimDocumentCollectionMapper.java]: Done

File: ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/domain/CCDClaimDocumentType.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (johnhosolirius) [reply ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/domain/CCDClaimDocumentType.java]: As above

File: ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/mapper/ClaimDocumentCollectionMapper.java (hunk 3 above)
Comment (johnhosolirius) [reply ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/mapper/ClaimDocumentCollectionMapper.java]: Done

File: ccd-adapter/src/test/java/uk/gov/hmcts/cmc/ccd/mapper/ClaimDocumentMapperTest.java (hunk 4 above)
Comment (johnhosolirius) [reply ccd-adapter/src/test/java/uk/gov/hmcts/cmc/ccd/mapper/ClaimDocumentMapperTest.java]: It doesn't "have" to be static. What would be the advantage of the alternative?

File: ccd-adapter/src/test/java/uk/gov/hmcts/cmc/ccd/mapper/ClaimDocumentMapperTest.java (hunk 6 above)
Comment (johnhosolirius) [reply ccd-adapter/src/test/java/uk/gov/hmcts/cmc/ccd/mapper/ClaimDocumentMapperTest.java]: done

File: src/main/java/uk/gov/hmcts/cmc/claimstore/events/DocumentUploadHandler.java (hunk 8 above)
Comment (johnhosolirius) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/events/DocumentUploadHandler.java]: Will be addressed in a PR from Srineelesh. So will resolve there.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/events/ccd/CCDCaseHandler.java (hunk 9 above)
Comment (johnhosolirius) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/events/ccd/CCDCaseHandler.java]: done

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/ClaimService.java (hunk 11 above)
Comment (johnhosolirius) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/ClaimService.java]: It's because the event would otherwise get created before the document is generate.

File: gradle/wrapper/gradle-wrapper.properties (hunk 7 above)
Comment (johnhosolirius) [reply gradle/wrapper/gradle-wrapper.properties]: It won't build for me unless I up the version number

File: src/main/java/uk/gov/hmcts/cmc/claimstore/repositories/CCDCaseRepository.java (hunk 10 above)
Comment (johnhosolirius) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/repositories/CCDCaseRepository.java]: done

File: gradle/wrapper/gradle-wrapper.properties (hunk 7 above)
Comment (karthikkannankainos) [reply gradle/wrapper/gradle-wrapper.properties]: @ckalista pls check if this is okay. and should it be part of this PR

File: ccd-adapter/src/test/java/uk/gov/hmcts/cmc/ccd/mapper/ClaimDocumentMapperTest.java (hunk 4 above)
Comment (karthikkannankainos) [reply ccd-adapter/src/test/java/uk/gov/hmcts/cmc/ccd/mapper/ClaimDocumentMapperTest.java]: static initializes variable at the start of the application. assume like lazy loading. So if not static it wont initialize it unless needed or used.

File: ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/mapper/CaseEventMapper.java (hunk 1 above)
Comment (johnhosolirius) [reply ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/mapper/CaseEventMapper.java]: As discussed. For the document type to event mapping:
1. It was the default before we added these mappings, so it's simply doing what it did before.
2. By throw an exception, we're effectively adding functionality and an additional step for support to deal with
3. We don't want the document upload to fail just because we can't find a mapping for the event type
4. In 3 the case worker will still be able to view the document in CCD.
5. The document will still get upload, albeit possibly not the correct event type. I think this is okay because it probably just means that no specific event type is defined for the document type and it would just be added as a future requirement

File: gradle/wrapper/gradle-wrapper.properties (hunk 7 above)
Comment (johnhosolirius) [reply gradle/wrapper/gradle-wrapper.properties]: Okay. I've removed it

File: src/test/java/uk/gov/hmcts/cmc/claimstore/services/ccd/CoreCaseDataServiceFailureTest.java [hunk 14]
```diff
@@ -564,3 +564,4 @@ public void savePaidInFullSubmitEventFailure() {
         service.savePaidInFull(claim.getId(), paidInFull, AUTHORISATION);
     }
 }
+ClaimDocumentMapperTest.java:
```
Comment (karthikkannankainos) [inline src/test/java/uk/gov/hmcts/cmc/claimstore/services/ccd/CoreCaseDataServiceFailureTest.java]: what is this?

Comment (johnhosolirius) [reply src/test/java/uk/gov/hmcts/cmc/claimstore/services/ccd/CoreCaseDataServiceFailureTest.java]: My bad. I can type faster than Intellij can respond :-)

I actually fixed and pushed about an ago. Sorry about.

File: gradle/wrapper/gradle-wrapper.properties (hunk 7 above)
Comment (ckalista) [reply gradle/wrapper/gradle-wrapper.properties]: Thank you guys for tying up the loose ends
Although fairly confident this PR won’t create any issue as it is. But since we have discussed this now I would be more inclined to bump up the version number in a separate PR if you don’t mind. 👍

File: src/test/java/uk/gov/hmcts/cmc/claimstore/services/ccd/CoreCaseDataServiceTest.java [hunk 15]
```diff
@@ -253,7 +253,8 @@ public void linkSealedClaimDocumentShouldReturnCaseDetails() {
 
         CaseDetails caseDetails = service.saveClaimDocuments(AUTHORISATION,
             SampleClaim.CLAIM_ID,
-            claim.getClaimDocumentCollection().orElse(new ClaimDocumentCollection()));
+            claim.getClaimDocumentCollection().orElse(new ClaimDocumentCollection()),
```
Comment (andrewwa-kainos) [inline src/test/java/uk/gov/hmcts/cmc/claimstore/services/ccd/CoreCaseDataServiceTest.java]: Suggest to use orElseGet(ClaimDocumentCollection::new) to avoid unnecessary object instantiation.

File: src/test/java/uk/gov/hmcts/cmc/claimstore/services/ccd/CoreCaseDataServiceFailureTest.java [hunk 16]
```diff
@@ -246,7 +247,8 @@ public void linkSealedClaimDocumentFailure() {
 
         service.saveClaimDocuments(AUTHORISATION,
             SampleClaim.CLAIM_ID,
-            claim.getClaimDocumentCollection().orElse(new ClaimDocumentCollection()));
+            claim.getClaimDocumentCollection().orElse(new ClaimDocumentCollection()),
```
[the remark this answers is not in the document]
Comment (johnhosolirius) [reply src/test/java/uk/gov/hmcts/cmc/claimstore/services/ccd/CoreCaseDataServiceFailureTest.java]: Agreed that although the comment is fair and good practice, it okay to leave it.

File: src/test/java/uk/gov/hmcts/cmc/claimstore/services/ccd/CoreCaseDataServiceTest.java (hunk 15 above)
Comment (johnhosolirius) [reply src/test/java/uk/gov/hmcts/cmc/claimstore/services/ccd/CoreCaseDataServiceTest.java]: Agreed that although the comment is fair and good practice, it okay to leave it.

File: domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/ClaimDocument.java [hunk 17]
```diff
@@ -7,7 +7,7 @@
 import java.time.LocalDateTime;
 
 @Getter
-public class ClaimDocument {
+public class ClaimDocument extends CollectionId {
```
Comment (markry-kainos) [inline domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/ClaimDocument.java]: not specifically relevant to your PR but I don't like `CollectionId` name 😸, should this be something `CCDCollectionItem` or `CollectionItemWithId` - its not an ID, it's something with an ID.
